File: api-net/Controllers/InvoiceController.cs
using System.Text.Json;
using Liftonom.Api.Data;
using Liftonom.Api.Models;
using Liftonom.Api.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace Liftonom.Api.Controllers;

[ApiController]
[Authorize]
[Route("api/v1/invoices")]
public class InvoiceController(AppDbContext db, LedgerService ledger, PdfService pdf) : ControllerBase
{
    public record InvoiceDto(long CustomerId, string? Type, DateOnly? IssueDate, DateOnly? DueDate,
        List<LineItem>? Items, decimal? TaxRate, decimal? Discount, string? Notes);
    public record PayDto(decimal Amount, long CashboxId, string PaymentMethod);

    [HttpGet]
    public async Task<IActionResult> Index([FromQuery] string? status, [FromQuery(Name = "customer_id")] long? customerId,
        [FromQuery(Name = "per_page")] int perPage = 25, [FromQuery] int page = 1)
    {
        var q = db.Invoices.AsQueryable();
        if (!string.IsNullOrEmpty(status)) q = q.Where(i => i.Status == status);
        if (customerId is { } c) q = q.Where(i => i.CustomerId == c);
        var projected = q.OrderByDescending(i => i.Id).Select(i => new
        {
            i.Id, i.InvoiceNumber, i.Status, i.Total, i.PaidAmount, i.IssueDate, i.DueDate,
            Customer = i.CustomerId == null ? null : new { i.Customer!.Name },
        });
        return Ok(await projected.ToPagedAsync(page, perPage));
    }

    [HttpPost]
    public async Task<IActionResult> Store(InvoiceDto dto)
    {
        var t = DocumentTotals.Compute(dto.Items, dto.TaxRate ?? 20, dto.Discount ?? 0);
        var now = DateTime.UtcNow;
        var inv = new Invoice
        {
            TenantId = db.CurrentTenantId!.Value, CustomerId = dto.CustomerId, Type = dto.Type ?? "e-archive",
            Status = "draft", IssueDate = dto.IssueDate ?? DateOnly.FromDateTime(DateTime.UtcNow), DueDate = dto.DueDate,
            Items = JsonSerializer.Serialize(dto.Items ?? []), Subtotal = t.Subtotal, TaxRate = t.TaxRate,
            TaxAmount = t.TaxAmount, Discount = t.Discount, Total = t.Total, Notes = dto.Notes,
            CreatedBy = long.Parse(User.FindFirst("uid")!.Value), CreatedAt = now, UpdatedAt = now,
        };
        db.Invoices.Add(inv);
        await db.SaveChangesAsync();
        inv.InvoiceNumber = $"FTR-{inv.Id:D6}";
        await db.SaveChangesAsync();
        return StatusCode(201, inv);
    }

    [HttpPost("from-quote/{quoteId:long}")]
    public async Task<IActionResult> FromQuote(long quoteId)
    {
        var q = await db.Quotes.FirstOrDefaultAsync(x => x.Id == quoteId) ?? throw new ApiException(404, "Teklif bulunamadı.");
        var now = DateTime.UtcNow;
        var inv = new Invoice
        {
            TenantId = db.CurrentTenantId!.Value, CustomerId = q.CustomerId, Status = "draft",
            IssueDate = DateOnly.FromDateTime(DateTime.UtcNow), Items = q.Items, Subtotal = q.Subtotal,
            TaxRate = q.TaxRate, TaxAmount = q.TaxAmount, Discount = q.Discount, Total = q.Total,
            SourceType = "quote", SourceId = q.Id, CreatedBy = long.Parse(User.FindFirst("uid")!.Value),
            CreatedAt = now, UpdatedAt = now,
        };
        db.Invoices.Add(inv);
        await db.SaveChangesAsync();
        inv.InvoiceNumber = $"FTR-{inv.Id:D6}";
        await db.SaveChangesAsync();
        return StatusCode(201, inv);
    }

    [HttpGet("{id:long}")]
    public async Task<IActionResult> Show(long id) =>
        Ok(await db.Invoices.Include(i => i.Customer).FirstOrDefaultAsync(i => i.Id == id) ?? throw new ApiException(404, "Fatura bulunamadı."));

    [HttpPost("{id:long}/send")]
    public async Task<IActionResult> Send(long id)
    {
        var inv = await db.Invoices.FirstOrDefaultAsync(x => x.Id == id) ?? throw new ApiException(404, "Fatura bulunamadı.");
        inv.Status = "sent"; inv.UpdatedAt = DateTime.UtcNow;
        await db.SaveChangesAsync();
        return Ok(inv);
    }

    [HttpPost("{id:long}/pay")]
    public async Task<IActionResult> Pay(long id, PayDto dto)
    {
        var inv = await db.Invoices.FirstOrDefaultAsync(x => x.Id == id) ?? throw new ApiException(404, "Fatura bulunamadı.");
        var uid = long.Parse(User.FindFirst("uid")!.Value);
        await ledger.CollectAsync(inv.CustomerId!.Value, dto.Amount, dto.PaymentMethod, dto.CashboxId,
            $"Fatura {inv.InvoiceNumber} ödemesi", uid);
        inv.PaidAmount += dto.Amount;
        if (inv.PaidAmount >= (inv.Total ?? 0)) inv.Status = "paid";
        inv.UpdatedAt = DateTime.UtcNow;
        await db.SaveChangesAsync();
        return Ok(inv);
    }

    [HttpGet("{id:long}/pdf")]
    public async Task<IActionResult> Pdf(long id)
    {
        var inv = await db.Invoices.Include(i => i.Customer).FirstOrDefaultAsync(i => i.Id == id)
            ?? throw new ApiException(404, "Fatura bulunamadı.");
        var tenant = await db.Tenants.FindAsync(db.CurrentTenantId!.Value);
        var bytes = pdf.Generate("FATURA", inv.InvoiceNumber ?? $"#{inv.Id}", tenant!, inv.Customer?.Name ?? "-",
            inv.Items, inv.Subtotal ?? 0, inv.TaxRate, inv.TaxAmount ?? 0, inv.Discount, inv.Total ?? 0, inv.IssueDate);
        return File(bytes, "application/pdf", $"{inv.InvoiceNumber}.pdf");
    }

    [HttpDelete("{id:long}")]
    public async Task<IActionResult> Destroy(long id)
    {
        var inv = await db.Invoices.FirstOrDefaultAsync(x => x.Id == id) ?? throw new ApiException(404, "Fatura bulunamadı.");
        inv.DeletedAt = DateTime.UtcNow;
        await db.SaveChangesAsync();
        return Ok(new { message = "Fatura silindi." });
    }
}

File: api/app/Http/Controllers/Api/InvoiceController.php
<?php
// app/Http/Controllers/Api/InvoiceController.php

namespace App\Http\Controllers\Api;

use App\Http\Controllers\Controller;
use App\Models\Invoice;
use App\Models\Quote;
use App\Services\LedgerService;
use App\Support\DocumentTotals;
use Illuminate\Http\Request;

class InvoiceController extends Controller
{
    public function index(Request $request)
    {
        $q = Invoice::query()->with('customer:id,name');
        if ($request->filled('status')) {
            $q->where('status', $request->string('status'));
        }
        if ($request->filled('customer_id')) {
            $q->where('customer_id', $request->integer('customer_id'));
        }

        return $q->orderByDesc('id')->paginate(min((int) $request->integer('per_page', 25), 100));
    }

    public function store(Request $request)
    {
        $data = $request->validate([
            'customer_id' => ['required', 'integer', 'exists:customers,id'],
            'type'        => ['nullable', 'in:e-fatura,e-archive'],
            'issue_date'  => ['nullable', 'date'],
            'due_date'    => ['nullable', 'date'],
            'items'       => ['nullable', 'array'],
            'tax_rate'    => ['nullable', 'numeric'],
            'discount'    => ['nullable', 'numeric'],
            'notes'       => ['nullable', 'string'],
        ]);
        $totals = DocumentTotals::compute($data['items'] ?? [], (float) ($data['tax_rate'] ?? 20), (float) ($data['discount'] ?? 0));
        $invoice = Invoice::create(array_merge($data, $totals, [
            'status'     => 'draft',
            'issue_date' => $data['issue_date'] ?? now()->toDateString(),
            'created_by' => $request->user()->id,
        ]));
        $invoice->update(['invoice_number' => 'FTR-' . str_pad((string) $invoice->id, 6, '0', STR_PAD_LEFT)]);

        return response()->json($invoice->fresh()->load('customer:id,name'), 201);
    }

    /** Tekliften fatura oluştur. */
    public function fromQuote(Request $request, Quote $quote)
    {
        $invoice = Invoice::create([
            'customer_id' => $quote->customer_id,
            'status'      => 'draft',
            'issue_date'  => now()->toDateString(),
            'items'       => $quote->items,
            'subtotal'    => $quote->subtotal,
            'tax_rate'    => $quote->tax_rate,
            'tax_amount'  => $quote->tax_amount,
            'discount'    => $quote->discount,
            'total'       => $quote->total,
            'source_type' => 'quote',
            'source_id'   => $quote->id,
            'created_by'  => $request->user()->id,
        ]);
        $invoice->update(['invoice_number' => 'FTR-' . str_pad((string) $invoice->id, 6, '0', STR_PAD_LEFT)]);

        return response()->json($invoice->fresh(), 201);
    }

    public function show(Invoice $invoice)
    {
        return $invoice->load('customer:id,name');
    }

    public function update(Request $request, Invoice $invoice)
    {
        $data = $request->validate([
            'due_date' => ['nullable', 'date'],
            'items'    => ['nullable', 'array'],
            'tax_rate' => ['nullable', 'numeric'],
            'discount' => ['nullable', 'numeric'],
            'notes'    => ['nullable', 'string'],
        ]);
        if (isset($data['items'])) {
            $data = array_merge($data, DocumentTotals::compute($data['items'], (float) ($data['tax_rate'] ?? $invoice->tax_rate), (float) ($data['discount'] ?? $invoice->discount)));
        }
        $invoice->update($data);

        return $invoice->fresh();
    }

    public function destroy(Invoice $invoice)
    {
        $invoice->delete();

        return response()->json(['message' => 'Fatura silindi.']);
    }

    public function send(Invoice $invoice)
    {
        $invoice->update(['status' => 'sent']);

        return $invoice->fresh();
    }

    /** Fatura ödemesi — ledger'a tahsilat işler. */
    public function pay(Request $request, Invoice $invoice, LedgerService $ledger)
    {
        $data = $request->validate([
            'amount'         => ['required', 'numeric', 'min:0.01'],
            'cashbox_id'     => ['required', 'integer', 'exists:cashboxes,id'],
            'payment_method' => ['required', 'in:cash,card,transfer,check'],
        ]);

        $ledger->collect(
            $invoice->customer_id,
            (float) $data['amount'],
            $data['payment_method'],
            $data['cashbox_id'],
            "Fatura {$invoice->invoice_number} ödemesi",
            $request->user()->id,
        );

        $invoice->paid_amount = (float) $invoice->paid_amount + (float) $data['amount'];
        if ((float) $invoice->paid_amount >= (float) $invoice->total) {
            $invoice->status = 'paid';
        }
        $invoice->save();

        return $invoice->fresh();
    }
}
